### PR TITLE
PorousFlow nodal material action

### LINKED ISSUE #11716: PorousFlow: automatically add correct version of materials as necessary
## Rationale

Due to the lumping and full upwinding in the PorousFlow module, some kernels need materials with the properties evaluated at the quadrature points (we call these qp materials), some need materials evaluated using nodal variables (we call these nodal materials), and some need both.

Each PorousFlow material has a parameter `at_nodes` which we set to `true` if it is a nodal material, and `false` if it is a qp material. This means that a typical input file contains several duplicate material entries, with the only difference being the value of the `at_nodes` parameter.

Eg:

````
[Materials]
  [./temperature]
    type = PorousFlowTemperature
    at_nodes = true
  [../]
  [./temperature_qp]
    type = PorousFlowTemperature
    at_nodes = false
  [../]
  [./ppss]
    type = PorousFlow2PhasePS
    at_nodes = true
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./ppss_qp]
    type = PorousFlow2PhasePS
    at_nodes = false
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./porosity]
    type = PorousFlowPorosityConst
    at_nodes = true
    porosity = 0.1
  [../]
  [./permeability]
    type = PorousFlowPermeabilityConst
    at_nodes = false
    permeability = '1e-15 0 0 0 1e-15 0 0 0 1e-15'
  [../]
  [./relperm_water]
    type = PorousFlowRelativePermeabilityCorey
    at_nodes = true
    n = 1
    phase = 0
  [../]
  [./relperm_gas]
    type = PorousFlowRelativePermeabilityCorey
    at_nodes = true
    n = 1
    phase = 1
  [../]
[]
````
etc.

This can be confusing for our users. How do they know whether they need a qp material, a nodal material, or both?  At the moment, it is probably by trial and error, looking through the source code, using one of the tests as a guide etc.

We have some actions available to help, but the user still has to specify a number of materials that the actions don't set up.

We know which materials each kernel requires, so we should be able to add the correct versions as required, making it easier for users.

## Description

The action system could be used to add the correct version (qp or nodal) of each material, so that each material only needs to be added once. The above input would then look like
````
[Materials]
  [./temperature]
    type = PorousFlowTemperature
  [../]
  [./ppss]
    type = PorousFlow2PhasePS
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./porosity]
    type = PorousFlowPorosityConst
    porosity = 0.1
  [../]
  [./permeability]
    type = PorousFlowPermeabilityConst
    permeability = '1e-15 0 0 0 1e-15 0 0 0 1e-15'
  [../]
  [./relperm_water]
    type = PorousFlowRelativePermeabilityCorey
    n = 1
    phase = 0
  [../]
  [./relperm_gas]
    type = PorousFlowRelativePermeabilityCorey
    n = 1
    phase = 1
  [../]
[]
````

## Impact

Make it easier for users to set up PorousFlow simulations

### PR DESCRIPTION
**Pull request by cpgr**
The basic idea of this is a new `meta_action`, `PorousFlowAddMaterialAction`, that checks each of the `PorousFlowMaterials` to see what value the `at_nodes` parameter should be set to, rather than relying on the user to work it out, and follows on from previous efforts at making PorousFlow easier to use. 

In an input file, the user can now simply add materials like so:
```
[Materials]
  [./temperature]
    type = PorousFlowTemperature
  [../]
  [./ppss]
    type = PorousFlow2PhasePS
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./porosity]
    type = PorousFlowPorosityConst
    porosity = 0.1
  [../]
[]
```

This action then checks the kernels and actions that are added in the input file to see whether we need `at_nodes = true`, `at_nodes = false`, or even both. It then either modifies the `at_nodes` parameter before the material is added if only one version is needed, and adds another to the action warehouse if both are required. This means that it can replace duplicate material entries in the input file where only the `at_nodes` parameter is different. In this example, the above input may replace the following:
```
[Materials]
  [./temperature]
    type = PorousFlowTemperature
  [../]
  [./temperature_nodal]
    type = PorousFlowTemperature 
    at_nodes = true
  [../]
  [./ppss]
    type = PorousFlow2PhasePS
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./ppss_nodal]
    type = PorousFlow2PhasePS
    at_nodes = true
    phase0_porepressure = ppwater
    phase1_saturation = sgas
    capillary_pressure = pc
  [../]
  [./porosity]
    type = PorousFlowPorosityConst
    porosity = 0.1
  [../]
  [./porosity_nodal]
    type = PorousFlowPorosityConst
    at_nodes = true
    porosity = 0.1
  [../]
[]
```

This is backwards compatible with the current way of doing things, so shouldn't break anyones current input files (there is a test added to check this), but I have simplified the tests where possible to remove most of these duplicate material entries (where the only difference is the `at_nodes` parameter).

Closes #11716

### CODE REVIEW

**Comment by cpgr** [comment]
I fixed the problems with the documentation in #11817

File: framework/include/utils/DependencyResolver.h [hunk 1]
```diff
@@ -419,6 +428,21 @@ DependencyResolver<T>::dependsOn(const std::vector<T> & keys, const T & value)
   return false;
 }
 
+template <typename T>
+const std::vector<T> &
+DependencyResolver<T>::getValues(const T & key)
+{
+  _values_vector.clear();
+
+  std::pair<typename std::multimap<T, T>::iterator, typename std::multimap<T, T>::iterator> ret;
```
Comment (permcody) [inline framework/include/utils/DependencyResolver.h]: Ha! Maybe `auto` next time? 😄

**Review by permcody — APPROVED** [review]
This really cleans up input files. Glad you got all this working!

File: framework/include/utils/DependencyResolver.h (hunk 1 above)
Comment (cpgr) [reply framework/include/utils/DependencyResolver.h]: Once I started I couldn't stop

**Comment by cpgr** [comment]
Thanks @permcody

**Pull request merged**